### PR TITLE
Changes to support Sync replication

### PR DESCRIPTION
**Pull request by rajendraindukuri**
# PR Submission checklist

# GitHub Issues
| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/1443 |

# Common PR Checklist:

- [x] Have you made sure that the code compiles?
- [x] Have you commented your code, particularly in hard-to-understand areas
- [ ] Did you run tests in a real Kubernetes cluster?
- [x] Have you maintained backward compatibility

## Description of your changes:
- Added Zero to the list of RPOEnum values
- Added unit tests for RPOEnum and CreateReplicationRule use case to reflect sync replication
- Added int tests for sync replication until volume group creation

## Unit test results:
image

image

## Integration test result for Sync Replication:
=== RUN   TestReplicationSuiteSync
2024/09/16 18:28:27 REQUEST: GET /api/rest/remote_system?select=id%2Cname%2Cdescription%2Cserial_number%2Cmanagement_address%2Cdata_connection_state%2Ccapabilities HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:27 acquire a lock
2024/09/16 18:28:28 RESPONSE: HTTP/1.1 200 OK Content-Length: 323 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  [{"id":"8fb61bf0-5216-471b-9c08-ca4845d97b4b","name":"RT-D7416","description":"","serial_number":"PS30ddf4e1d754","management_address":"10.x.y.z","data_connection_state":"OK","capabilities":["Asynchronous_Block_Replication","Asynchronous_Vvol_Replication","Block_Metro_Active_Active","Synchronous_Block_Replication"]}]
2024/09/16 18:28:28 release a lock
2024/09/16 18:28:28 REQUEST: GET /api/rest/remote_system/8fb61bf0-5216-471b-9c08-ca4845d97b4b?select=id%2Cname%2Cdescription%2Cserial_number%2Cmanagement_address%2Cdata_connection_state%2Ccapabilities HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:28 acquire a lock
2024/09/16 18:28:28 RESPONSE: HTTP/1.1 200 OK Content-Length: 321 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  {"id":"8fb61bf0-5216-471b-9c08-ca4845d97b4b","name":"RT-D7416","description":"","serial_number":"PS30ddf4e1d754","management_address":"10.x.y.z","data_connection_state":"OK","capabilities":["Asynchronous_Block_Replication","Asynchronous_Vvol_Replication","Block_Metro_Active_Active","Synchronous_Block_Replication"]}
2024/09/16 18:28:28 release a lock
2024/09/16 18:28:28 REQUEST: POST /api/rest/replication_rule HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"name":"intcsiDCHqbHmW-ruletst","rpo":"Zero","remote_system_id":"8fb61bf0-5216-471b-9c08-ca4845d97b4b"}
2024/09/16 18:28:28 acquire a lock
2024/09/16 18:28:28 RESPONSE: HTTP/1.1 201 Created Transfer-Encoding: chunked Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  2d {"id":"03e9b6a7-7ce4-4013-9da5-b02269555574"} 0
2024/09/16 18:28:28 release a lock
2024/09/16 18:28:28 REQUEST: POST /api/rest/policy HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"name":"intcsiDCHqbHmW-pptst","replication_rule_ids":["03e9b6a7-7ce4-4013-9da5-b02269555574"],"snapshot_rule_ids":null}
2024/09/16 18:28:28 acquire a lock
2024/09/16 18:28:29 RESPONSE: HTTP/1.1 201 Created Transfer-Encoding: chunked Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  2d {"id":"ad1f2bb4-d423-4217-bddb-340b87b2a655"} 0
2024/09/16 18:28:29 release a lock
2024/09/16 18:28:29 Session management is enabled.
2024/09/16 18:28:29 REQUEST: GET /api/rest/login_session HTTP/1.1 Host: 10.x.y.z Authorization: ******
2024/09/16 18:28:29 acquire a lock
2024/09/16 18:28:30 RESPONSE: HTTP/1.1 200 OK Transfer-Encoding: chunked Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff X-Content-Type-Options: nosniff  2f [{"id":"d5b9d89e-6c3f-46a6-a6ca-b24485a29b97"}] 0
2024/09/16 18:28:30 release a lock
=== RUN   TestReplicationSuiteSync/TestReplicationSync
2024/09/16 18:28:30 REQUEST: GET /api/rest/remote_system/8fb61bf0-5216-471b-9c08-ca4845d97b4b?select=id%2Cname%2Cdescription%2Cserial_number%2Cmanagement_address%2Cdata_connection_state%2Ccapabilities HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:30 acquire a lock
2024/09/16 18:28:30 RESPONSE: HTTP/1.1 200 OK Content-Length: 321 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  {"id":"8fb61bf0-5216-471b-9c08-ca4845d97b4b","name":"RT-D7416","description":"","serial_number":"PS30ddf4e1d754","management_address":"10.x.y.z","data_connection_state":"OK","capabilities":["Asynchronous_Block_Replication","Asynchronous_Vvol_Replication","Block_Metro_Active_Active","Synchronous_Block_Replication"]}
2024/09/16 18:28:30 release a lock
2024/09/16 18:28:30 REQUEST: POST /api/rest/volume_group HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"name":"intcsiDCHqbHmW-vgtst","protection_policy_id":"ad1f2bb4-d423-4217-bddb-340b87b2a655","is_write_order_consistent":true}
2024/09/16 18:28:30 acquire a lock
2024/09/16 18:28:30 RESPONSE: HTTP/1.1 201 Created Transfer-Encoding: chunked Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  2d {"id":"d313047f-4e1c-4fa6-a0ed-71f86d327372"} 0
2024/09/16 18:28:30 release a lock
2024/09/16 18:28:30 REQUEST: POST /api/rest/volume HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"name":"intcsiDCHqbHmW-voltst","size":1048576,"volume_group_id":"d313047f-4e1c-4fa6-a0ed-71f86d327372"}
2024/09/16 18:28:30 acquire a lock
2024/09/16 18:28:31 RESPONSE: HTTP/1.1 201 Created Transfer-Encoding: chunked Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  2d {"id":"8255968a-b4c0-4efa-b982-e614d9bc9852"} 0
2024/09/16 18:28:31 release a lock
2024/09/16 18:28:31 REQUEST: GET /api/rest/volume/8255968a-b4c0-4efa-b982-e614d9bc9852?select=volume_groups%28id%2Cname%2Cprotection_policy_id%29 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:31 acquire a lock
2024/09/16 18:28:31 RESPONSE: HTTP/1.1 200 OK Content-Length: 157 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  {"volume_groups":[{"id":"d313047f-4e1c-4fa6-a0ed-71f86d327372","name":"intcsiDCHqbHmW-vgtst","protection_policy_id":"ad1f2bb4-d423-4217-bddb-340b87b2a655"}]}
2024/09/16 18:28:31 release a lock
time="2024-09-16T18:28:31+05:30" level=info msg="{[{d313047f-4e1c-4fa6-a0ed-71f86d327372 intcsiDCHqbHmW-vgtst  ad1f2bb4-d423-4217-bddb-340b87b2a655 false [] {      false false   [] [] [] [] [] [] []}  false false  false    {   } [] { }}]}"
--- PASS: TestReplicationSuiteSync/TestReplicationSync (1.53s)
2024/09/16 18:28:31 REQUEST: GET /api/rest/volume_group/d313047f-4e1c-4fa6-a0ed-71f86d327372?select=%2A%2Cvolumes%28%2A%29%2Cprotection_policy%28%2A%29%2Cprotection_data%2Clocation_history%2Cmigration_session%28%2A%29%2Cvolume.volume_group_membership%28id%2Cname%2Cprotection_policy_id%2Cstate%2Cprotection_data%29 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:31 acquire a lock
2024/09/16 18:28:31 RESPONSE: HTTP/1.1 200 OK Content-Length: 2872 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  {"id":"d313047f-4e1c-4fa6-a0ed-71f86d327372","name":"intcsiDCHqbHmW-vgtst","description":null,"creation_timestamp":"2024-09-16T12:58:31.607986+00:00","is_protectable":true,"protection_policy_id":"ad1f2bb4-d423-4217-bddb-340b87b2a655","qos_performance_policy_id":null,"migration_session_id":null,"is_write_order_consistent":true,"placement_rule":"Same_Appliance","type":"Primary","is_replication_destination":false,"metro_replication_session_id":null,"is_host_access_available":true,"protection_data":{"family_id": "d313047f-4e1c-4fa6-a0ed-71f86d327372", "is_secure": false, "parent_id": null, "source_id": null
, "creator_type": "User", "copy_signature": null, "source_timestamp": null, "creator_type_l10n": "User", "is_app_consistent": false, "created_by_rule_id": null, "created_by_rule_name": null, "expiration_timestamp": null},"is_importing":false,"location_history":null,"type_l10n":"Primary","volumes":[{"id":"8255968a-b4c0-4efa-b982-e614d9bc9852","name":"intcsiDCHqbHmW-voltst","description":"","type":"Primary","wwn":"naa.68ccf09800b9024392887cf11cf0dce6","nsid":12939,"nguid":"nguid.b9024392887cf11c8ccf096800f0dce6","appliance_id":"A1","state":"Ready","size":1048576,"logical_used":null,"node_affinity":"System_Select_At_Attach","creation_timestamp":"2024-09-16T12:58:32.073852+00:00","protection_policy_id":null,"performance_policy_id":"default_medium","qos_performance_policy_id":null,"is_replication_destination":false,"migration_session_id":null,"metro_replication_session_id":null,"is_host_access_available":true,"protection_data":{"family_id": "8255968a-b4c0-4efa-b982-e614d9bc9852", "is_secure": false, "parent_id": n
ull, "source_id": null, "creator_type": "User", "copy_signature": null, "source_timestamp": null, "creator_type_l10n": "User", "is_app_consistent": false, "created_by_rule_id": null, "created_by_rule_name": null, "expiration_timestamp": null},"location_history":null,"app_type":null,"app_type_other":null,"type_l10n":"Primary","state_l10n":"Ready","node_affinity_l10n":"System Select At Attach","app_type_l10n":null}],"protection_policy":{"id":"ad1f2bb4-d423-4217-bddb-340b87b2a655","name":"intcsiDCHqbHmW-pptst","description":null,"type":"Protection","managed_by":"User","managed_by_id":null,"is_read_only":false,"is_replica":false,"file_io_limit_rule_id":null,"type_l10n":"Protection","managed_by_l10n":"Managed by the end user"},"migration_session":null,"volume":[{"id":"8255968a-b4c0-4efa-b982-e614d9bc9852","name":"intcsiDCHqbHmW-voltst","protection_policy_id":null,"state":"Ready","protection_data":{"family_id": "8255968a-b4c0-4efa-b982-e614d9bc9852", "is_secure": false, "parent_id": null, "source_id": null, "creato
r_type": "User", "copy_signature": null, "source_timestamp": null, "creator_type_l10n": "User", "is_app_consistent": false, "created_by_rule_id": null, "created_by_rule_name": null, "expiration_timestamp": null}}]}
2024/09/16 18:28:31 release a lock
2024/09/16 18:28:31 REQUEST: GET /api/rest/policy?name=eq.intcsiDCHqbHmW-pptst&select=%2A%2Creplication_rules%28%2A%29%2Csnapshot_rules%28%2A%29%2Cvirtual_machines%28%2A%29%2Cfile_systems%28%2A%29%2Cperformance_rules%28%2A%29%2Cvolumes%28%2A%29%2Cvolume_groups%28%2A%29&type=eq.Protection HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:31 acquire a lock
2024/09/16 18:28:32 RESPONSE: HTTP/1.1 200 OK Content-Length: 1642 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  [{"id":"ad1f2bb4-d423-4217-bddb-340b87b2a655","name":"intcsiDCHqbHmW-pptst","description":null,"type":"Protection","managed_by":"User","managed_by_id":null,"is_read_only":false,"is_replica":false,"file_io_limit_rule_id":null,"type_l10n":"Protection","managed_by_l10n":"Managed by the end user","replication_rules":[{"id":"03e9b6a7-7ce4-4013-9da5-b02269555574","name":"intcsiDCHqbHmW-ruletst","rpo":"Zero","remote_system_id":"8fb61bf0-5216-471b-9c08-ca4845d97b4b","is_replica":false,"is_read_only":false,"alert_threshold":0,"managed_by":"User","managed_by_id":null,"rpo_l10n":"Zero","managed_by_l10n":"Managed b
y the end user"}],"snapshot_rules":[],"virtual_machines":[],"file_systems":[],"performance_rules":[],"volumes":[],"volume_groups":[{"id":"d313047f-4e1c-4fa6-a0ed-71f86d327372","name":"intcsiDCHqbHmW-vgtst","description":null,"creation_timestamp":"2024-09-16T12:58:31.607986+00:00","is_protectable":true,"protection_policy_id":"ad1f2bb4-d423-4217-bddb-340b87b2a655","qos_performance_policy_id":null,"migration_session_id":null,"is_write_order_consistent":true,"placement_rule":"Same_Appliance","type":"Primary","is_replication_destination":false,"metro_replication_session_id":null,"is_host_access_available":true,"protection_data":{"family_id": "d313047f-4e1c-4fa6-a0ed-71f86d327372", "is_secure": false, "parent_id": null, "source_id": null, "creator_type": "User", "copy_signature": null, "source_timestamp": null, "creator_type_l10n": "User", "is_app_consistent": false, "created_by_rule_id": null, "created_by_rule_name": null, "expiration_timestamp": null},"is_importing":false,"location_history":null,"type_l10n":"Prim
ary"}]}]
2024/09/16 18:28:32 release a lock
2024/09/16 18:28:32 REQUEST: GET /api/rest/replication_rule?name=eq.intcsiDCHqbHmW-ruletst&select=id%2Cname%2Crpo%2Cremote_system_id%2Cpolicies HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:32 acquire a lock
2024/09/16 18:28:32 RESPONSE: HTTP/1.1 200 OK Content-Length: 209 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  [{"id":"03e9b6a7-7ce4-4013-9da5-b02269555574","name":"intcsiDCHqbHmW-ruletst","rpo":"Zero","remote_system_id":"8fb61bf0-5216-471b-9c08-ca4845d97b4b","policies":[{"id":"ad1f2bb4-d423-4217-bddb-340b87b2a655"}]}]
2024/09/16 18:28:32 release a lock
2024/09/16 18:28:32 REQUEST: PATCH /api/rest/volume_group/d313047f-4e1c-4fa6-a0ed-71f86d327372 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"protection_policy_id":"","description":""}
2024/09/16 18:28:32 acquire a lock
2024/09/16 18:28:33 RESPONSE: HTTP/1.1 204 No Content Content-Length: 0 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:33 release a lock
2024/09/16 18:28:33 REQUEST: POST /api/rest/volume_group/d313047f-4e1c-4fa6-a0ed-71f86d327372/remove_members HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"volume_ids":["8255968a-b4c0-4efa-b982-e614d9bc9852"]}
2024/09/16 18:28:33 acquire a lock
2024/09/16 18:28:33 RESPONSE: HTTP/1.1 204 No Content Content-Length: 0 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:33 release a lock
2024/09/16 18:28:33 REQUEST: PATCH /api/rest/volume/8255968a-b4c0-4efa-b982-e614d9bc9852 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******  {"protection_policy_id":"","description":""}
2024/09/16 18:28:33 acquire a lock
2024/09/16 18:28:34 RESPONSE: HTTP/1.1 204 No Content Content-Length: 0 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:34 release a lock
2024/09/16 18:28:34 REQUEST: DELETE /api/rest/policy/ad1f2bb4-d423-4217-bddb-340b87b2a655 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:34 acquire a lock
2024/09/16 18:28:34 RESPONSE: HTTP/1.1 204 No Content Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:34 release a lock
2024/09/16 18:28:34 REQUEST: DELETE /api/rest/replication_rule/03e9b6a7-7ce4-4013-9da5-b02269555574 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:34 acquire a lock
2024/09/16 18:28:35 RESPONSE: HTTP/1.1 204 No Content Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:35 release a lock
2024/09/16 18:28:35 REQUEST: DELETE /api/rest/volume_group/d313047f-4e1c-4fa6-a0ed-71f86d327372 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:35 acquire a lock
2024/09/16 18:28:35 RESPONSE: HTTP/1.1 204 No Content Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:35 release a lock
2024/09/16 18:28:35 REQUEST: GET /api/rest/volume_group?name=eq.intcsiDCHqbHmW-vgtst&select=%2A%2Cvolumes%28%2A%29%2Cprotection_policy%28%2A%29%2Cprotection_data%2Clocation_history%2Cmigration_session%28%2A%29 HTTP/1.1 Host: 10.x.y.z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:35 acquire a lock
2024/09/16 18:28:35 RESPONSE: HTTP/1.1 200 OK Content-Length: 2 Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff  []
2024/09/16 18:28:35 release a lock
time="2024-09-16T18:28:35+05:30" level=info
2024/09/16 18:28:35 REQUEST: DELETE /api/rest/volume_group HTTP/1.1 Host: 10.x.y.z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:35 acquire a lock
2024/09/16 18:28:36 RESPONSE: HTTP/1.1 405 Method Not Allowed Content-Length: 140 Content-Type: application/json Dell-Emc-Token: ****** Set-Cookie: auth_cookie=; Max-Age=0; Expires=Mon, 16 Sep 2024 12:58:36 GMT Strict-Transport-Security: max-age=31536000; includeSubDomains  {"messages":[{"severity":"Error","code":"0xE04040030002","message_l10n":"Request method is not supported: DELETE.","arguments":["DELETE"]}]}
2024/09/16 18:28:36 release a lock
2024/09/16 18:28:36 REQUEST: DELETE /api/rest/volume/8255968a-b4c0-4efa-b982-e614d9bc9852 HTTP/1.1 Host: 10.X.Y.Z Authorization: ****** Dell-Emc-Token: ******
2024/09/16 18:28:36 acquire a lock
2024/09/16 18:28:36 RESPONSE: HTTP/1.1 204 No Content Cache-Control: no-cache Cache-Control: no-store Cache-Control: must-revalidate Cache-Control: max-age=0 Content-Language: en-US Content-Type: application/json Dell-Emc-Token: ****** Expires: -1 Set-Cookie: auth_cookie=******; Path=/; Secure; HTTPOnly Strict-Transport-Security: max-age=31536000; includeSubDomains X-Content-Type-Options: nosniff
2024/09/16 18:28:36 release a lock
--- PASS: TestReplicationSuiteSync (8.99s)

### CODE REVIEW

**Review by lukeatdell — CHANGES_REQUESTED** [review]
Please also update or add integration tests under /inttests to test the changes.

File: replication_test.go [hunk 1]
```diff
@@ -73,6 +73,24 @@ func TestClientIMPL_CreateReplicationRule(t *testing.T) {
 	assert.Equal(t, volID, resp.ID)
 }
```
Comment (santhoshatdell) [inline replication_test.go:74]: Copyright year needs update.

**Review by santhoshatdell — APPROVED** [review]
Please address Luke's comment.

File: replication_test.go (hunk 1 above)
Comment (rajendraindukuri) [reply replication_test.go:74]: done

File: replication_types.go [hunk 2]
```diff
@@ -60,6 +61,7 @@ type ReplicationRuleCreate struct {
 	// Name of the replication rule.
 	Name string `json:"name"`
 	// Recovery point objective (RPO), which is the acceptable amount of data, measured in units of time, that may be lost in case of a failure.
+	// If RPO is Zero, it indicates the replication_type is 'sync'.
```
[the remark this answers is not in the document]
Comment (rajendraindukuri) [reply replication_types.go:64]: done

**Comment by rajendraindukuri** [comment]
@lukeatdell Added int tests for Sync Replication until volume group creation. This can be enhanced to add Replication session related tests post the changes are implemented. Thanks.

**Comment by rajendraindukuri** [comment]
> Please address Luke's comment.

Done

File: inttests/replication_sync_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
+/*
+ *
+ * Copyright © 2021-2024 Dell Inc. or its subsidiaries. All Rights Reserved.
```
Comment (AkshaySainiDell) [inline inttests/replication_sync_test.go:3]:
```suggestion
 * Copyright © 2024 Dell Inc. or its subsidiaries. All Rights Reserved.
```

File: replication_types_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/*
+ *
+ * Copyright © 2021-2024 Dell Inc. or its subsidiaries. All Rights Reserved.
```
Comment (AkshaySainiDell) [inline replication_types_test.go:3]:
```suggestion
 * Copyright © 2024 Dell Inc. or its subsidiaries. All Rights Reserved.
```

**Review by AkshaySainiDell — CHANGES_REQUESTED** [review]

**Review by lukeatdell — DISMISSED** [review]
Thanks for adding the tests!

**Review by AkshaySainiDell — APPROVED** [review]

**Review by HarishH-DELL — APPROVED** [review]

**Pull request merged**